eval/src/vespa/eval/instruction/detect_type.h

```diff
@@ -0,0 +1,75 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <typeindex>
+#include <array>
+
+#pragma once
+
+namespace vespalib::eval::instruction {
+
+/*
+ * Utilities for detecting implementation class by comparing
+ * typeindex(typeid(T)); for now these are local to this
+ * namespace, but we can consider moving them to a more
+ * common place (probably vespalib) if we see more use-cases.
+ */
+
+/**
+ * Recognize a (const) instance of type T.  This is cheaper than
+ * dynamic_cast, but requires the object to be exactly of class T.
+ * Returns a pointer to the object as T if recognized, nullptr
+ * otherwise.
+ **/
+template<typename T, typename U>
+const T *
+recognize_by_type_index(const U & object)
+{
+    if (std::type_index(typeid(object)) == std::type_index(typeid(T))) {
+        return static_cast<const T *>(&object);
+    }
+    return nullptr;
+}
+
+/**
+ * Packs N recognized values into one object, used as return value
+ * from detect_type<T>.
+ * 
+ * Use all_converted() or the equivalent bool cast operator to check
+ * if all objects were recognized.  After this check is successful use
+ * get<0>(), get<1>() etc to get a reference to the objects.
+ **/
+template<typename T, size_t N>
+class RecognizedValues
+{
+private:
+    std::array<const T *, N> _pointers;
+public:
+    RecognizedValues(std::array<const T *, N> && pointers)
+        : _pointers(std::move(pointers))
+    {}
+    bool all_converted() const {
+        for (auto p : _pointers) {
+            if (p == nullptr) return false;            
+        }
+        return true;
+    }
+    operator bool() const { return all_converted(); }
+    template<size_t idx> const T& get() const {
```

Comment:
what about operator() ?
indexes = detect_types<FastValueIndex>(a, b)
indexes<0>()
indexes<1>()

Reply:
that would have been prettier, but it doesn't compile at all; according to stackoverflow you have to write
`indexes.operator()<0>();`

Reply:
ok, was worth a try :-)